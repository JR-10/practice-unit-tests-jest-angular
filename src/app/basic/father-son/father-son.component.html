<h5>Hijo</h5>

<!-- para formatear codigo JSON -->
<code>{{ client | json }}</code>

<br><br>

<button *ngIf="client" class="btn btn-outline-primary" (click)="onChangeId(5)" data-test="btn-id">Cambiar ID</button>

<button *ngIf="client" class="btn btn-outline-danger" (click)="onDelete()" data-test="btn-delete">Borrar CLiente</button>
