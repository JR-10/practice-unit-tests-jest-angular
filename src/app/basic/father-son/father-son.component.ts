import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { Client } from '../interfaces';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.scss']
})
export class FatherSonComponent implements OnInit {

  // Capturar el valor del padre con @Input
  @Input() client?: Client

  // emitir evento
  @Output() onDeleteClient = new EventEmitter();

  // emitir objeto
  @Output() onClientUpdated = new EventEmitter<Client>();

  constructor() { }

  ngOnInit(): void {
  }


  /**
   * @description Metodo para disparar el evento
   */
  onDelete(){
    this.client = undefined; // eliminar el cliente
    this.onDeleteClient.emit(); // indicarle al padre que se elimino el cliente
  }


  /**
   * @description Metodo para realizar el cambio del id al cliente
   * @param id
   */
  onChangeId(id: number){
    if(!this.client) return; // validacion si el cliente no existe que no haga nada es

    // si el cliente existe
    // this.client.id = newId; // no usar ya que actualiza  tanto padre como hijo ya que pasa pro referencia

    this.client = { ...this.client, id}; // romper la referencia por medio del spread de ese cliente

    // emision del evento
    this.onClientUpdated.emit({...this.client});

  }

}
