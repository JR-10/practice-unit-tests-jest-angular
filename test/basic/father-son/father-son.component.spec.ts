import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement; // para unit test con HTML

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    compiled = fixture.nativeElement;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe de hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });


  test('no deben de aparecer botones si no hay cliente', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  });

  test('deben de aparecer 2 botones si hay cliente', () => {
    component.client = {id: 1, name: 'Juan'};
    fixture.detectChanges();
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  });

  test('Si hay cliente, hacer match con el snapshot ', () => {
    component.client = {id: 1, name: 'Pedro'};
    fixture.detectChanges();
    expect(compiled).toMatchSnapshot();
  });

  test('debe de emitir onDeleteClient con el boton de eliminar', () => {
    component.client = {id: 1, name: 'Pedro'};
    fixture.detectChanges();
    // espiar en el componente
    jest.spyOn(component.onDeleteClient, 'emit'); // funcion especial de jest para espiar
    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    btnDelete?.dispatchEvent(new Event('click')); // simular un eventop click
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });

  test('debe de emitir onClientUpdated con el boton Cambiar ID', () => {
    component.client = {id: 1, name: 'Pedro'};
    fixture.detectChanges();
    // espiar en el componente
    jest.spyOn(component.onClientUpdated, 'emit'); // funcion especial de jest para espiar
    const btnChangeId = compiled.querySelector('[data-test=btn-id]');
    btnChangeId?.dispatchEvent(new Event('click')); // simular un eventop click
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({id: 5, name: 'Pedro'});
  });

  test('deben de emitir onChangeClient con el  ID especificado SI hay un cliente', () => {
    jest.spyOn(component.onClientUpdated, 'emit');
    component.onChangeId(10);
    expect(component.onClientUpdated.emit).not.toHaveBeenCalled();

    component.client = {id: 1, name: 'Pedro'};
    fixture.detectChanges();
    component.onChangeId(10);
    expect(component.onClientUpdated.emit).not.toHaveBeenCalledWith({id: 5, name: 'Pedro'});

  });

});
