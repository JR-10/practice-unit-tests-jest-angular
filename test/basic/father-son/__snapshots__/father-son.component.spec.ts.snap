// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent Si hay cliente, hacer match con el snapshot  1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "Pedro"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-id"
  >
    Cambiar ID
  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
    Borrar CLiente
  </button>
  
</div>
`;

exports[`FatherSonComponent debe de hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;
